workflow-runtime/src/commonTest/kotlin/com/squareup/workflow1/RenderWorkflowInTest.kt

```diff
@@ -377,85 +341,168 @@ class RenderWorkflowInTest {
-      val trigger = Channel<String>()
-      val workflow = Workflow.stateful<String, String, String>(
-        initialState = "initial",
-        render = { renderState ->
-          runningWorker(
-            trigger.consumeAsFlow()
-              .asWorker()
-          ) {
-            action("") {
-              state = it
-              setOutput(it)
+      runTest(UnconfinedTestDispatcher()) {
+        val trigger = Channel<String>()
+        val workflow = Workflow.stateful<String, String, String>(
+          initialState = "initial",
+          render = { renderState ->
+            runningWorker(
+              trigger.receiveAsFlow()
+                .asWorker()
+            ) {
+              action("") {
+                state = it
+                setOutput(it)
+              }
             }
+            renderState
           }
-          renderState
+        )
+
+        val receivedOutputs = mutableListOf<String>()
+        lateinit var renderings: StateFlow<RenderingAndSnapshot<String>>
+        renderings = renderWorkflowIn(
+          workflow = workflow,
+          scope = backgroundScope,
+          props = MutableStateFlow(Unit),
+          runtimeConfig = runtimeConfig,
+          workflowTracer = workflowTracer,
+        ) { it: String ->
+          receivedOutputs += it
+          // The value of the updated rendering has already been set by the time onOutput is
+          // called
+          assertEquals(it, renderings.value.rendering)
         }
-      )
-
-      val emittedRenderings = mutableListOf<String>()
-      val receivedOutputs = mutableListOf<String>()
-      val renderings = renderWorkflowIn(
-        workflow = workflow,
-        scope = testScope,
-        props = MutableStateFlow(Unit),
-        runtimeConfig = runtimeConfig,
-        workflowTracer = workflowTracer,
-      ) { it: String ->
-        receivedOutputs += it
-        assertTrue(emittedRenderings.contains(it))
+        assertTrue(receivedOutputs.isEmpty())
+
+        assertTrue(trigger.trySend("foo").isSuccess)
+        assertEquals(listOf("foo"), receivedOutputs)
       }
-      assertTrue(receivedOutputs.isEmpty())
+    }
+  }
 
-      val scope = CoroutineScope(Unconfined)
-      scope.launch {
-        renderings.collect { rendering: RenderingAndSnapshot<String> ->
-          emittedRenderings += rendering.rendering
+  /**
+   * A different form of [onOutput_called_after_rendering_emitted]. Here we launch the workflow
+   * runtime on its own [TestScope] with a [StandardTestDispatcher], which will be paused until
+   * told to advance.
+   *
+   * We *collect* emitted renderings on the [UnconfinedTestDispatcher] of the [runTest].
+   * The point here is that when the runtime sets the value on the StateFlow - a non-suspending
+   * operation - and then it calls [onOutput] - a suspending operation - the [onOutput] handler
+   * will not be immediately dispatched (it is waiting for dispatch from the scheduler), but the
```

Comment:
> the [onOutput] handler will not be immediately dispatched (it is waiting for dispatch from the scheduler)

The workflow runtime calls `onOutput` directly, it's just a function, so the dispatcher is not involved at all. The reason this timing is the way it is is because the runtime sets the MSF's `value` property before calling `onOutput`. Or am I misreading this explanation?

Reply:
`onOutput` is a `suspend fun`, so it will go through dispatch. It's a good point though that perhaps we don't need it to be a `suspend fun`

Reply:
I'm thinking we made `onOutput` a `suspend fun` because we wanted it to be able to cancel the runtime's scope? cc @rjrjr if you remember.

Reply:
No idea. I doubt we did it by accident.